blockchain/spam_throttler.go

```diff
@@ -0,0 +1,304 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package blockchain
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/rcrowley/go-metrics"
+
+	"github.com/klaytn/klaytn/blockchain/types"
+	"github.com/klaytn/klaytn/common"
+)
+
+// TODO-Klaytn: move these variables into TxPool when BlockChain struct contains a TxPool interface
+// spamThrottler need to be accessed by both of TxPool and BlockChain.
+var (
+	spamThrottler   *throttler = nil
+	spamThrottlerMu            = new(sync.RWMutex)
+)
+
+var (
+	thresholdGauge           = metrics.NewRegisteredGauge("txpool/throttler/threshold", nil)
+	candidateSizeGauge       = metrics.NewRegisteredGauge("txpool/throttler/candidate/size", nil)
+	throttledSizeGauge       = metrics.NewRegisteredGauge("txpool/throttler/throttled/size", nil)
+	allowedSizeGauge         = metrics.NewRegisteredGauge("txpool/throttler/allowed/size", nil)
+	throttlerUpdateTimeGauge = metrics.NewRegisteredGauge("txpool/throttler/update/time", nil)
+	throttlerDropCount       = metrics.NewRegisteredCounter("txpool/throttler/dropped/count", nil)
+)
+
+type throttler struct {
+	config *ThrottlerConfig
+
+	candidates map[common.Address]int  // throttle candidates with spam weight. Not for concurrent use
+	throttled  map[common.Address]int  // throttled addresses with throttle time. Requires mu.lock for concurrent use
+	allowed    map[common.Address]bool // white listed addresses. Requires mu.lock for concurrent use
+	mu         *sync.RWMutex           // mutex for throttled and allowed
+
+	threshold  int
+	throttleCh chan *types.Transaction
+	quitCh     chan struct{}
+}
+
+type ThrottlerConfig struct {
+	ActivateTxPoolSize uint `json:"activate_tx_pool_size"`
+	TargetFailRatio    uint `json:"target_fail_ratio"`
+	ThrottleTPS        uint `json:"throttle_tps"`
+	MaxCandidates      uint `json:"max_candidates"`
+
+	IncreaseWeight      int `json:"increase_weight"`
+	DecreaseWeight      int `json:"decrease_weight"`
+	InitialThreshold    int `json:"initial_threshold"`
+	MinimumThreshold    int `json:"minimum_threshold"`
+	ThresholdAdjustment int `json:"threshold_adjustment"`
+	ThrottleSeconds     int `json:"throttle_seconds"`
+}
+
+var DefaultSpamThrottlerConfig = &ThrottlerConfig{
+	ActivateTxPoolSize: 1000,
+	TargetFailRatio:    20,
+	ThrottleTPS:        100,   // len(throttleCh) = ThrottleTPS * 5. 32KB * 100 * 5 = 16MB
+	MaxCandidates:      10000, // (20 + 4)B * 10000 = 240KB
+
+	IncreaseWeight:      5,
+	DecreaseWeight:      1,
+	InitialThreshold:    500,
+	MinimumThreshold:    100,
+	ThresholdAdjustment: 5,
+	ThrottleSeconds:     300,
+}
+
+func GetSpamThrottler() *throttler {
+	spamThrottlerMu.RLock()
+	t := spamThrottler
+	spamThrottlerMu.RUnlock()
+	return t
+}
+
+func validateConfig(conf *ThrottlerConfig) error {
+	if conf == nil {
+		return errors.New("nil ThrottlerConfig")
+	}
+	if conf.TargetFailRatio > 100 {
```

Comment:
```suggestion
	if conf.TargetFailRatio > 100 || conf.TargetFailRatio < 0 {
```

Based on error message, I think is should be handle when the given `conf.TargetFailRatio` is under zero.

Reply:
I didn't put the condition because the type of TargetFailRatio is `uint`, but I will consider adding the condition for make sure later

Reply:
@aidan-kwon 
My miss..!! I didn't consider that variable is `uint`. Thanks for your reply